### PR TITLE
Automatic thesis anonymization after 5-year retention period

### PR DESCRIPTION
**Pull request by krusche**
## Summary

- Add automatic thesis anonymization after the 5-year legal retention period expires, including scheduled batch processing and email notifications to research group heads
- Add admin-only manual delete (anonymize) for individual theses with a warning system for non-terminal state and active retention periods
- Add architecture test ensuring no Jackson 2.x databind/core imports in production code (project uses Jackson 3.x)

### Thesis Anonymization (automatic)
- New `ThesisAnonymizationService` with batch anonymization (`anonymizeExpiredTheses`) and notification logic (`sendAnonymizationNotifications`)
- `RetentionUtils` computes retention expiry (5 years from thesis end date or last state change)
- Liquibase migration adds `anonymized_at` and `anonymization_notified_at` columns, plus email template
- Anonymized theses preserve structural data (title, type, grade, dates) for statistics while removing all personal data (files, comments, assessments, feedback, role assignments)

### Admin Manual Delete (anonymize)
- `GET /v2/theses/{id}/anonymize/warnings` — returns warnings (non-terminal state, active retention, already anonymized)
- `DELETE /v2/theses/{id}/anonymize` — admin-only endpoint to anonymize a single thesis
- UI: "Delete Thesis" button + confirmation modal with orange warning alerts in thesis config section
- Reuses the same `anonymizeThesis()` method as the batch process

### Frontend
- Anonymized thesis detail page shows informational banner
- Thesis DTO includes `anonymized` and `anonymizedAt` fields
- Anonymized theses hide personal data sections (roles, comments, files, etc.)
- Admin page includes "Run Thesis Anonymization" button for on-demand batch execution

### Other
- Remove `EntityManager` usage from `UserDeletionService` (reorder operations to avoid Hibernate stale-state instead)
- Move seed data to end of Liquibase changelog so it can reference all columns from prior migrations
- Jackson 2.x architecture test prevents accidental use of old databind/core imports

## Test plan

- [x] 714 server tests pass (`./gradlew test`)
- [x] 118 E2E tests pass (`./execute-e2e-local.sh`)
- [x] Checkstyle passes (`./gradlew checkstyleMain checkstyleTest`)
- [x] Client lint passes (`npx eslint src/`)
- [x] Client type check passes (`npx tsc --noEmit`)

## Summary by CodeRabbit

* **New Features**
  * Admins can run thesis anonymization from Admin page; per-thesis anonymize action with warnings/confirmation; anonymization banners and retention dates shown on anonymized theses; scheduled notification/reminder support added.

* **Bug Fixes**
  * Prevented runtime rendering errors by defaulting missing students/advisors/supervisors to empty lists.

* **Tests**
  * Added extensive e2e, integration, and unit tests covering anonymization, notifications, retention, deletion, and access controls.

* **Documentation**
  * Expanded developer docs with Data Management Tests and anonymization coverage.

### CODE REVIEW

**Review by Claudia-Anthropica — CHANGES_REQUESTED** [review]
@krusche Solid feature — the anonymization design is well-thought-out, especially the retention computation, file collection before deletion, and the idempotent batch processing. However, CI is red: `doesNotReNotifyAlreadyNotifiedThesis` fails due to a timestamp precision mismatch between Java and PostgreSQL. See inline comment for the fix.

File: server/src/test/java/de/tum/cit/aet/thesis/service/ThesisAnonymizationServiceTest.java [hunk 1]
```diff
@@ -0,0 +1,413 @@
...
+		@Test
+		void doesNotReNotifyAlreadyNotifiedThesis() throws Exception {
+			createTestEmailTemplate("THESIS_ANONYMIZATION_REMINDER");
+
+			Instant createdAt = Instant.now().minus(2600, ChronoUnit.DAYS);
+			Instant endDate = Instant.now().minus(2400, ChronoUnit.DAYS);
+			Thesis thesis = createTestThesisWithChildren(ThesisState.FINISHED, createdAt, endDate);
+
+			// Mark as already notified
+			thesis.setAnonymizationNotifiedAt(Instant.now().minus(10, ChronoUnit.DAYS));
```
[the remark this answers is not in the document]
Comment (Claudia-Anthropica) [reply server/src/test/java/de/tum/cit/aet/thesis/service/ThesisAnonymizationServiceTest.java]: @krusche This is still unfixed — the `truncatedTo(ChronoUnit.MICROS)` is still missing on line 402. The CI test `doesNotReNotifyAlreadyNotifiedThesis` will keep failing until this is addressed.

File: client/e2e/thesis-delete.spec.ts [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+    await expect(alert).toBeVisible({ timeout: 5_000 })
+    await expect(alert.getByText(/WRITING/i)).toBeVisible()
+    // Should also show retention warning since thesis is only 30 days old
+    await expect(alert.getByText(/retention period/i)).toBeVisible()
+
+    // Confirm delete despite warnings
+    await dialog.getByRole('button', { name: 'Delete Thesis' }).click()
+
+    await expect(page).toHaveURL(/\/theses(?:\?|$)/, { timeout: 15_000 })
+  })
```
[the remark this answers is not in the document]
Comment (krusche) [reply client/e2e/thesis-delete.spec.ts]: The E2E tests use dedicated seed data (theses 10-12) separate from main test data, and the tests run in serial order. Rerun safety is ensured by the test infrastructure resetting the database between test suite runs.

File: client/src/pages/ThesisPage/components/ThesisConfigSection/ThesisConfigSection.tsx [hunk 3]
```diff
@@ -239,149 +297,196 @@ const ThesisConfigSection = () => {
...
+                  {...form.getInputProps('advisors')}
+                />
+                <UserMultiSelect
+                  required={true}
+                  disabled={!access.advisor}
+                  label='Examiner'
+                  groups={['supervisor']}
+                  initialUsers={thesis.supervisors}
+                  maxValues={1}
+                  {...form.getInputProps('supervisors')}
```
[the remark this answers is not in the document]
Comment (krusche) [reply client/src/pages/ThesisPage/components/ThesisConfigSection/ThesisConfigSection.tsx]: Fixed in f7ac351. Added `?? []` to all three UserMultiSelect initialUsers props.

File: server/src/main/java/de/tum/cit/aet/thesis/dto/ThesisAnonymizationResultDto.java [hunk 4]
```diff
@@ -0,0 +1,4 @@
+package de.tum.cit.aet.thesis.dto;
+
+public record ThesisAnonymizationResultDto(int anonymizedTheses) {
+}
```
[the remark this answers is not in the document]
Comment (krusche) [reply server/src/main/java/de/tum/cit/aet/thesis/dto/ThesisAnonymizationResultDto.java]: Fixed in f7ac351. Added `@JsonInclude(JsonInclude.Include.NON_EMPTY)` annotation.

File: server/src/main/java/de/tum/cit/aet/thesis/repository/ThesisProposalRepository.java [hunk 5]
```diff
@@ -2,11 +2,21 @@
...
 import java.util.UUID;
 
 @Repository
 public interface ThesisProposalRepository extends JpaRepository<ThesisProposal, UUID> {
+	@Modifying
+	@Transactional
+	void deleteAllByThesisId(UUID thesisId);
 
+	@Query("SELECT p.proposalFilename FROM ThesisProposal p WHERE p.thesis.id = :thesisId")
+	List<String> findFilenamesByThesisId(@Param("thesisId") UUID thesisId);
```
[the remark this answers is not in the document]
Comment (krusche) [reply server/src/main/java/de/tum/cit/aet/thesis/repository/ThesisProposalRepository.java]: Fixed in f7ac351. Added `AND p.proposalFilename IS NOT NULL` to the query.

File: server/src/main/java/de/tum/cit/aet/thesis/service/ThesisAnonymizationService.java [hunk 6]
```diff
@@ -0,0 +1,273 @@
...
+		thesisPresentationInviteRepository.deleteAllByPresentationThesisId(thesisId);
+		thesisPresentationRepository.deleteAllByThesisId(thesisId);
+		thesisCommentRepository.deleteAllByThesisId(thesisId);
+		thesisFileRepository.deleteAllByThesisId(thesisId);
+		thesisProposalRepository.deleteAllByThesisId(thesisId);
+		thesisAssessmentRepository.deleteAllByThesisId(thesisId);
+		thesisFeedbackRepository.deleteAllByThesisId(thesisId);
+		thesisStateChangeRepository.deleteAllByThesisId(thesisId);
+		thesisRoleRepository.deleteAllByThesisId(thesisId);
+
```
[the remark this answers is not in the document]
Comment (krusche) [reply server/src/main/java/de/tum/cit/aet/thesis/service/ThesisAnonymizationService.java]: Fixed in f7ac351. Wrapped child record deletions in a try/catch that resets `anonymizedAt` to null on failure, keeping the thesis eligible for retry.

File: server/src/main/java/de/tum/cit/aet/thesis/service/ThesisService.java [hunk 7]
```diff
@@ -130,6 +130,12 @@ private CurrentUserProvider currentUserProvider() {
 		return currentUserProviderProvider.getObject();
 	}
 
+	private void requireNotAnonymized(Thesis thesis) {
+		if (thesis.isAnonymized()) {
+			throw new ResourceInvalidParametersException("This thesis has been anonymized and cannot be modified.");
+		}
+	}
```
[the remark this answers is not in the document]
Comment (krusche) [reply server/src/main/java/de/tum/cit/aet/thesis/service/ThesisService.java]: Fixed in f7ac351. Added `requireNotAnonymized(thesis)` to all 7 remaining mutating methods: completeFeedback, deleteFeedback, requestChanges, deleteProposal, acceptProposal, deleteThesisFile, completeThesis.

File: server/src/test/java/de/tum/cit/aet/thesis/architecture/JacksonImportArchitectureTest.java [hunk 8]
```diff
@@ -0,0 +1,61 @@
+package de.tum.cit.aet.thesis.architecture;
+
+import static org.junit.jupiter.api.Assertions.assertTrue;
```
[the remark this answers is not in the document]
Comment (krusche) [reply server/src/test/java/de/tum/cit/aet/thesis/architecture/JacksonImportArchitectureTest.java]: Fixed in f7ac351. Switched from JUnit `assertTrue` to AssertJ `assertThat(...).isEmpty()`.

File: server/src/test/java/de/tum/cit/aet/thesis/architecture/JacksonImportArchitectureTest.java [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+	private static final Path MAIN_SRC = Paths.get("src/main/java");
+
+	/**
+	 * Pattern matching forbidden Jackson 2.x imports.
+	 * Matches {@code com.fasterxml.jackson.core} and {@code com.fasterxml.jackson.databind}
+	 * but NOT {@code com.fasterxml.jackson.annotation} (shared with Jackson 3.x).
+	 */
+	private static final Pattern FORBIDDEN_IMPORT = Pattern.compile(
+			"^import\\s+com\\.fasterxml\\.jackson\\.(core|databind)\\b"
+	);
```
[the remark this answers is not in the document]
Comment (krusche) [reply server/src/test/java/de/tum/cit/aet/thesis/architecture/JacksonImportArchitectureTest.java]: Fixed in f7ac351. Updated regex to `^import\\s+(?:static\\s+)?com\\.fasterxml\\.jackson\\.(core|databind)\\b` to also catch static imports.

File: server/src/main/java/de/tum/cit/aet/thesis/service/ThesisAnonymizationService.java [hunk 10]
```diff
@@ -0,0 +1,285 @@
...
+			thesisAssessmentRepository.deleteAllByThesisId(thesisId);
+			thesisFeedbackRepository.deleteAllByThesisId(thesisId);
+			thesisStateChangeRepository.deleteAllByThesisId(thesisId);
+			thesisRoleRepository.deleteAllByThesisId(thesisId);
+		} catch (Exception ex) {
+			// Keep thesis eligible for retry if cleanup failed
+			thesis.setAnonymizedAt(null);
+			thesisRepository.save(thesis);
+			throw ex;
+		}
```
[the remark this answers is not in the document]
Comment (krusche) [reply server/src/main/java/de/tum/cit/aet/thesis/service/ThesisAnonymizationService.java]: Fixed in e7d3581. Moved file deletion to a `finally` block so pre-collected filenames are always cleaned up, even if DB child deletion partially fails.

**Review by Claudia-Anthropica — APPROVED** [review]
@krusche All previous feedback addressed, and the new E2E tests look solid — good coverage of modal interactions, role restrictions, and edge cases. Nice work.

**Review by Claudia-Anthropica — APPROVED** [review]
@krusche All previous feedback addressed, all threads resolved. The latest commit cleanly removes PII from log messages and adds the missing E2E assertion. Looks good to merge.
